api-server/src/server/boot/challenge.js

```diff
@@ -82,6 +90,340 @@ export default async function bootChallenge(app, done) {
+  const { body } = req;
+
+  // Ensure `body` is an array:
+  if (!Array.isArray(body) || !body.length) {
+    return res.status(400).json({
+      type: 'error',
+      message: 'Invalid request format. Expected `body` to be an array.'
+    });
+  }
+  for (const challenge of body) {
+    const { id, challengeType } = challenge;
+
+    // Ensure `id` and `challengeType` exist on `challenge`
+    if (typeof id !== 'string' || typeof challengeType !== 'number') {
+      return res.status(400).json({
+        type: 'error',
+        message:
+          'Invalid request format. Expected `id` and `challengeType` to be present.'
+      });
+    }
+
+    const foundChallenge = allChallenges.find(chal => chal.id === id);
+    const isChallengeValid =
+      !!foundChallenge &&
+      ![
+        challengeTypes.frontEndProject,
+        challengeTypes.backEndProject,
+        challengeTypes.pythonProject,
+        challengeTypes.codeAllyCert,
+        challengeTypes.multifileCertProject
+      ].includes(challengeType);
+
+    if (!isChallengeValid) {
+      return res.status(400).json({
+        type: 'error',
+        message: 'Invalid challenge submission.'
+      });
+    }
+  }
+  next();
+}
+
+export function validateProject(req, res, next) {
+  const { body } = req;
+
+  // Ensure `body` is an object
+  if (!body || typeof body !== 'object') {
+    return res.status(400).json({
+      type: 'error',
+      message: 'Invalid request format. Expected `body` to be an object.'
+    });
+  }
+
+  const { id } = body;
+
+  // Find the structure, based on the `id` of the project
+  const challenge = allChallenges.find(chal => chal.id === id);
+
+  // Ensure `body` matches the expected structure
+  const validBody = ensureObjectContainsAllProperties(
+    body,
+    expectedProjectStructures[body.challengeType]
+  );
+  const isProjectValid =
+    !!challenge &&
+    [
+      challengeTypes.frontEndProject,
+      challengeTypes.backEndProject,
+      challengeTypes.bonfire,
+      challengeTypes.pythonProject,
+      challengeTypes.codeAllyCert,
+      challengeTypes.multifileCertProject
+    ].includes(body.challengeType) &&
+    validBody;
+
+  if (!isProjectValid) {
+    return res.status(400).json({
+      type: 'error',
+      message: 'Invalid project submission.'
+    });
+  }
+  next();
+}
+
+/**
+ * Handles submissions for all challenges.
+ * @param {Request} req
+ * @param {Response} res
+ * @returns {Observable}
+ */
+export function challengesCompleted(req, res, next) {
+  // Challenges should only include `id` and `challengeType`
+  const { user, body } = req;
+
+  const submittedChallenges = [];
+  for (const challenge of body) {
+    const { id, challengeType } = challenge;
+
+    // NOTE: Consider what will happen if this comes with a batch
+    const completedDate = Date.now();
+
+    const completedChallenge = {
+      challengeType, // TODO: A little birdy told me we do not need to store `challengeType` for non-certs?
```

Comment:
```suggestion
```
Yep, we haven't been storing it:
https://github.com/freeCodeCamp/freeCodeCamp/blob/d5423745d963a97fa33dc4584ade8d3e4d126551/api-server/src/server/boot/challenge.js#L284-L288